examples/external-subgraph/rust/src/echo_co_processor.rs

```diff
@@ -0,0 +1,193 @@
+    fn web_endpoints(&self) -> MultiMap<ListenAddr, Endpoint> {
+        let web_endpoint =
+            Endpoint::from_router_service("/".to_string(), SimpleEndpoint {}.boxed());
+
+        let mut endpoints = MultiMap::new();
+        let socket_addr: SocketAddr = format!("127.0.0.1:{}", self.configuration.port)
+            .parse()
+            .unwrap();
+        endpoints.insert(ListenAddr::from(socket_addr), web_endpoint);
+
+        endpoints
+    }
+}
+
+// This is a simple server, that will do a couple of transforms to payloads and return it.
+// In real life you will implement this outside the router, in your favorite language.
+struct SimpleEndpoint {}
+
+impl Service<router::Request> for SimpleEndpoint {
+    type Response = router::Response;
+
+    type Error = BoxError;
+
+    type Future = BoxFuture<'static, router::ServiceResult>;
+
+    fn poll_ready(
+        &mut self,
+        _cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<Result<(), Self::Error>> {
+        std::task::Poll::Ready(Ok(()))
+    }
+
+    fn call(&mut self, req: router::Request) -> Self::Future {
+        tracing::info!("📞 received request");
+        tracing::info!("JSON context:");
+        tracing::info!("{}", serde_json::to_string_pretty(&req.context).unwrap());
+
+        let fut = async move {
+            let body = req.router_request.into_body();
+
+            let body = hyper::body::to_bytes(body).await.unwrap();
+
+            let mut json_body: serde_json::Value = serde_json::from_slice(&body).unwrap();
+            tracing::info!("✉️ got payload:");
+            tracing::info!("{}", serde_json::to_string_pretty(&json_body).unwrap());
+
+            // let's add an arbitrary header to the request
+            if let Some(headers) = json_body.get_mut("headers") {
+                if let Some(headers) = headers.as_object_mut() {
+                    headers.insert(
+                        "x-my-subgraph-api-key".to_string(),
+                        json! {["ThisIsATestApiKey"]}, // header values are arrays
+                    );
+                }
+            } else {
+                json_body.as_object_mut().map(|body| {
+                    body.insert(
+                        "headers".to_string(),
+                        json! {{
+                            "x-my-subgraph-api-key": ["ThisIsATestApiKey"] // header values are arrays
+                        }},
+                    )
+                });
+            };
+
+            // let's add a context key so that the subgraph_http_service displays the headers it's about to send!
+            if let Some(context) = json_body.get_mut("context") {
+                let context = context.get_mut("entries").unwrap(); // context always has entries.
+                if let Some(context) = context.as_object_mut() {
+                    context.insert(
+                        "apollo_telemetry::logging::display_headers".to_string(),
```

Comment:
no we should not use that kind of example

Reply:
maybe use `apollo_authentication::JWT::claims` instead? Since that's a use case we want. That can even be justified from a subgraph extension: the router validates the JWT, adds claims, then the first subgraph looks at the claims and add more of them depending on user roles